File: server/routers/places.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models.place import Place
from schemas.place import PlaceCreate, PlaceUpdate, PlaceRead

router = APIRouter()


@router.get("/trips/{trip_id}/places", response_model=list[PlaceRead])
async def list_places(trip_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(Place)
        .where(Place.trip_id == trip_id)
        .order_by(Place.day_group, Place.sort_order)
    )
    return result.scalars().all()


@router.post("/trips/{trip_id}/places", response_model=PlaceRead, status_code=201)
async def create_place(trip_id: int, body: PlaceCreate, db: AsyncSession = Depends(get_db)):
    max_order_result = await db.execute(
        select(func.coalesce(func.max(Place.sort_order), -1)).where(
            Place.trip_id == trip_id
        )
    )
    max_order = max_order_result.scalar()

    place = Place(
        trip_id=trip_id,
        name=body.name,
        address=body.address,
        lat=body.lat,
        lng=body.lng,
        place_id=body.place_id,
        day_group=body.day_group,
        sort_order=max_order + 1,
    )
    db.add(place)
    await db.commit()
    await db.refresh(place)
    return place


@router.put("/places/{place_id}", response_model=PlaceRead)
async def update_place(place_id: int, body: PlaceUpdate, db: AsyncSession = Depends(get_db)):
    place = await db.get(Place, place_id)
    if not place:
        raise HTTPException(status_code=404, detail="Place not found")

    if body.notes is not None:
        place.notes = body.notes
    if body.day_group is not None:
        place.day_group = body.day_group
    if body.sort_order is not None:
        place.sort_order = body.sort_order

    await db.commit()
    await db.refresh(place)
    return place


@router.delete("/places/{place_id}", status_code=204)
async def delete_place(place_id: int, db: AsyncSession = Depends(get_db)):
    place = await db.get(Place, place_id)
    if not place:
        raise HTTPException(status_code=404, detail="Place not found")
    await db.delete(place)
    await db.commit()

File: server/src/routes/places.ts
import { Router } from 'express';
import db from '../db.js';

const router = Router();

router.get('/trips/:tripId/places', (req, res) => {
  const places = db
    .prepare('SELECT * FROM places WHERE trip_id = ? ORDER BY day_group, sort_order')
    .all(req.params.tripId);
  res.json(places);
});

router.post('/trips/:tripId/places', (req, res) => {
  const { name, address, lat, lng, place_id, day_group } = req.body;
  if (!name || lat == null || lng == null) {
    res.status(400).json({ error: 'name, lat, and lng are required' });
    return;
  }

  const maxOrder = db
    .prepare('SELECT COALESCE(MAX(sort_order), -1) as max_order FROM places WHERE trip_id = ?')
    .get(req.params.tripId) as { max_order: number };

  const result = db
    .prepare(
      `INSERT INTO places (trip_id, name, address, lat, lng, place_id, day_group, sort_order)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      req.params.tripId,
      name,
      address || '',
      lat,
      lng,
      place_id || '',
      day_group || '',
      maxOrder.max_order + 1
    );

  const place = db.prepare('SELECT * FROM places WHERE id = ?').get(result.lastInsertRowid);
  res.status(201).json(place);
});

router.put('/places/:id', (req, res) => {
  const { notes, day_group, sort_order } = req.body;
  const existing = db.prepare('SELECT * FROM places WHERE id = ?').get(req.params.id) as Record<string, unknown> | undefined;
  if (!existing) {
    res.status(404).json({ error: 'place not found' });
    return;
  }

  db.prepare(
    `UPDATE places SET notes = ?, day_group = ?, sort_order = ? WHERE id = ?`
  ).run(
    notes ?? existing.notes,
    day_group ?? existing.day_group,
    sort_order ?? existing.sort_order,
    req.params.id
  );

  const place = db.prepare('SELECT * FROM places WHERE id = ?').get(req.params.id);
  res.json(place);
});

router.delete('/places/:id', (req, res) => {
  const result = db.prepare('DELETE FROM places WHERE id = ?').run(req.params.id);
  if (result.changes === 0) {
    res.status(404).json({ error: 'place not found' });
    return;
  }
  res.status(204).end();
});

export default router;
